### PR TITLE
WIP: Persistence (session, history, bookmarks)

### PR DESCRIPTION
**Pull request by Ambrevar**
Draft for now.

I can restore URLs on startup.  Hooray! :)

I've extracted the startup precedure to make it configurable by the user.
The session "store" and "restore" functions can be customized.

Before proceding, I need to clean up parts of the code:

- buffer-set-url is used in places, like following link hints, which bypasses the `set-url` function.  I believe there should be only one entry point for URL change.  I suggest renaming buffer-set-url to `%set-url` and using it only from `set-url`.
- We could add an option to `set-url` to not call `parse-url`, e.g. `:raw-url-p`.
- I want to delete new-window and new-buffer.
- I want to merge `make-buffers` into `make-buffer`.

Question:

- Should the session persistence functions be per window or per interface?  Considering a session persistence might save information about windows, maybe it would make more sense to store it in the remote-interface class.  But what about "private windows" then?

### CODE REVIEW

**Comment by Ambrevar** [comment]
I've just pushed some changes to serialize the complete `buffer` data structure.
cl-prevalence does an amazing job here, short of 2 edge cases:
- pathnames
- functions

I've fixed the pathname serialization.

For functions, well it's not easy to serialize a function :p  For now, I'm serializing named functions to `#'my-function-name`, and unnamed functions to `#'ignoring`, where `ignoring` is a function that takes any argument and does nothing.  (Is there already such a function in the standard?)
It's not very pretty, but it allows us to serialize many class slots which have a function value (e.g. `resource-query-function`). 

I'd like to report those ideas upstream, see if we can get them merged.

The great news is that I'm now able to restore buffers with their history and all.  There are still some rough edges.
For instance, shall we re-run the mode constructors when restoring the buffer?  The constructor could set some mode slots, so we'd rather not.  But on the other side, the constructor could also make some RPC calls (e.g. enable the proxy) which are required for the mode to be on at all.

My take is that constructors should _not_ change the state of Next beyond the RPC calls.
I'm not sure how we can enforce that though.

Thoughts?

**Comment by Ambrevar** [comment]
I gave this a little more thoughts, and I'm realizing one thing: session persistence is a hard problem and I don't know if anyone ever got it right.  Emacs `desktop-mode` does not work so well in my experience (so many edge cases) and I prefer to rely on `recentf` to quickly re-access previously opened buffers.

As for restoring the right modes for the right URL, I just opened the issue #376.  So this could simply implemented orthogonally.

To sum up, the only thing we would have to serialize and restore are the buffer _histories_.  Which is a simple tree of URLs, so nothing hard here.

Thoughts?

**Comment by Ambrevar** [comment]
TODO: Update doc (search engines have been moved).

**Comment by Ambrevar** [comment]
Ready to merge, after 1.3.2.

**Comment by Ambrevar** [comment]
After the merge, we need to tackle #376, #377 and #378.
There is also #79 that could be dealt with while we are at it.

**Comment by Ambrevar** [comment]
And #380.

**Comment by Ambrevar** [comment]
Merged with 1eae4c8fe0fa12e462cb35ae4498c1d5e377a30e.
